did_peer/src/peer_did/numalgos/traits.rs

```diff
@@ -0,0 +1,50 @@
+use did_doc::schema::did_doc::DidDocument;
+use did_doc_sov::extra_fields::ExtraFieldsSov;
+use did_parser::Did;
+
+use crate::{
+    error::DidPeerError,
+    peer_did::{peer_did::PeerDid, validate::validate},
+    peer_did_resolver::options::PublicKeyEncoding,
+};
+
+use super::numalgo3::Numalgo3;
+
+pub trait Numalgo: Sized + Default {
+    const NUMALGO_CHAR: char;
+
+    fn parse<T>(did: T) -> Result<PeerDid<Self>, DidPeerError>
+    where
+        Did: TryFrom<T>,
+        <Did as TryFrom<T>>::Error: Into<DidPeerError>,
+    {
+        let did: Did = did.try_into().map_err(Into::into)?;
+
+        let numalgo_char = did.id().chars().nth(0).ok_or_else(|| {
+            DidPeerError::DidValidationError(format!(
+                "Invalid did: unable to read numalgo character in did {}",
+                did.did()
+            ))
+        })?;
+
+        if numalgo_char != Self::NUMALGO_CHAR {
+            return Err(DidPeerError::InvalidNumalgoCharacter(numalgo_char));
+        }
+
+        validate(&did)?;
+
+        Ok(PeerDid::from_parts(did, Self::default()))
+    }
+}
+
+pub trait ResolvableNumalgo: Numalgo {
+    fn resolve(
+        &self,
+        did: &Did,
+        public_key_encoding: PublicKeyEncoding,
+    ) -> Result<DidDocument<ExtraFieldsSov>, DidPeerError>;
+}
+
+pub trait ToNumalgo3: Numalgo {
+    fn to_numalgo3(&self, did: &Did) -> Result<PeerDid<Numalgo3>, DidPeerError>;
+}
```

Comment:
Could you provide some info about this trait? From what I see it's implemented for `Numalgo2` and `Numalgo3`, taking `&self` by reference but never really using it and always returning `PeerDid<Numalgo3>`.

I'm curious because I personally find that:
1) The name is a bit misleading -> You're not transforming between `Numalgo` variants, you actually return a `PeerDid`.
2) The signature seems weird given the implementations of the trait.
3) The purpose of the trait is unclear. Given that all of these are unit types, this could just as easily be a marker trait used for the trait bound in `impl<N: ToNumalgo3> PeerDid<N>` and have the actual implementation happen there.

Reply:
Implementors of the trait can be converted to peer:did:3. This is true for peer:did:2 since it can be shortened to peer:did:3 and for peer:did:3 because it already is peer:did:3. In some cases you don't care which numalgo you have as long as it can be converted to 3, e.g. querying a database for a record stored by peer:did:3.

Reply:
Yeah, on a second look I got mixed up between the implementations and they behave differently so a marker trait won't suffice. My bad, sorry. 

I'd still drop the `&self` in the signature though. 

Sealing it is definitely to be done, but I wouldn't honestly worry about that too much, at least not right now. I skipped sealing some stuff in `messages` and the new `Connection` myself as I was more focused on the actual implementation than a consumer randomly implementing a trait 😅 .

Definitely not proud of that now that I think about it, so I will revisit it soon 😄 .

Reply:
Removed `&self` from signature of `to_numalgo3`.